### ISSUE TITLE
When converting a class to a callable, both `__new__` and `__init__` methods should contribute

### ISSUE THREAD

**Issue by erictraut** [microsoft/pyright#7684]
The newly-updated typing spec indicates that both the `__new__` and `__init__` method should contribute to the resulting callable type.

Here is the test case from the conformance test suite. Currently, pyright doesn't generate an error in all of the correct cases because it's not evaluating a union type.

```python
class Class3:
    """__new__ and __init__"""

    def __new__(cls, *args, **kwargs) -> Self: ...

    def __init__(self, x: int) -> None: ...


class Expected3_1(Protocol):
    def __call__(self, *args: Any, **kwargs: Any) -> Class3: ...


class Expected3_2(Protocol):
    def __call__(self, x: int) -> Class3: ...


r3 = accepts_callable(Class3)
reveal_type(r3)  # ``def (*args, **kwargs) -> Class3 | def (x: int) -> Class3``
t3_1: Expected3_1 | Expected3_2 = r3  # OK
t3_2: Expected3_1 = r3  # E
t3_3: Expected3_2 = r3  # E
assert_type(r3(3), Class3)
r3()  # E
r3(1, 2)  # E
```

**Comment by erictraut** [comment]
This is included in pyright 1.1.359, which I just published.
